### ISSUE TITLE
Roll out cURL 8.12.1

### ISSUE THREAD

**Issue by cmb69** [winlibs/cURL#25]
cURL 8.11.0 has been released, fixing [CVE-2024-9681](https://curl.se/docs/CVE-2024-9681.html). Given that is a low severity issue, it *might* not be necessary to update stable branches right away (should wait after GA at least). I've already pushed the update to master, and did quick testing as usual, and found that now [Websocket support](https://curl.se/docs/websocket.html) is enabled by default. Probably not a problem, since that seems to require special support in ext/curl; otherwise I'd be wary to roll it out to stable versions.

Note that [nghttp2 1.64.0](https://github.com/winlibs/nghttp2/releases/tag/nghttp2-1.64.0) is available to be built as prerequisite for the cURL update.

@nielsdos, any thoughts about the update?

**Comment by cmb69** [comment]
Ah forgot: if we roll 8.11.0 out with Websocket support, we need to apply the following patch to php-src:

````diff
 ext/curl/tests/check_win_config.phpt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/ext/curl/tests/check_win_config.phpt b/ext/curl/tests/check_win_config.phpt
index b3beb044a7..8330a95564 100644
--- a/ext/curl/tests/check_win_config.phpt
+++ b/ext/curl/tests/check_win_config.phpt
@@ -54,7 +54,7 @@
 ZSTD => No
 HSTS => Yes
 GSASL => No
-Protocols => dict, file, ftp, ftps, gopher, %r(gophers, )?%rhttp, https, imap, imaps, ldap, ldaps, %r(mqtt, )?%rpop3, pop3s, rtsp, scp, sftp, smb, smbs, smtp, smtps, telnet, tftp
+Protocols => dict, file, ftp, ftps, gopher, %r(gophers, )?%rhttp, https, imap, imaps, ldap, ldaps, %r(mqtt, )?%rpop3, pop3s, rtsp, scp, sftp, smb, smbs, smtp, smtps, telnet, tftp%r(, ws, wss)?%r
 Host => %s-pc-win32
 SSL Version => OpenSSL/%s
 ZLib Version => %s
````

**Comment by ndossche** [comment]
Let's wait until after GA and then make sure the next release uses the update.

> Ah forgot: if we roll 8.11.0 out with Websocket support, we need to apply the following patch to php-src:

This likely needs to happen anyway for the Linux users who receive the update via their distro.

**Comment by cmb69** [comment]
> Let's wait until after GA and then make sure the next release uses the update.

Fine. I'll keep an eye on it.

> This likely needs to happen anyway for the Linux users who receive the update via their distro.

The test is Windows only. :)

**Comment by ndossche** [comment]
> Fine. I'll keep an eye on it.

Thanks!

> The test is Windows only. :)

Ah oops, I missed that. EDIT: duh, it even says win in the title... :facepalm:

**Comment by cmb69** [comment]
I guess we want to wait for cURL 8.11.1: https://curl.se/mail/lib-2024-11/0019.html

**Comment by cmb69** [comment]
I've pushed cURL 8.11.1 (which fixes another [low severity vulnerability](https://curl.se/docs/CVE-2024-11053.html)) to master. [Test build](https://github.com/winlibs/winlib-builder/actions/runs/12286110139) showed no further issues.

I suggest to wait with rolling out until PHP GA's have been released (scheduled for Dec 19th), and then first push staging to stable (we're behind with this for a couple of months). Afterwards we can roll out new releases.

**Comment by cmb69** [comment]
Hmm, that fell through the cracks. :(

Anyway, [cURL 8.12.0 has now been released](https://curl.se/ch/8.12.0.html), fixing 3 low severity security issues. We should update to this version right away (well, after the stable PHP versions have been rolled out on 2025-02-13). Besides the test case fix we would possibly have needed to fix (https://github.com/winlibs/cURL/issues/25#issuecomment-2460582125), we now also need https://github.com/php/php-src/pull/17709 (or something like that).

I've pushed and tagged the update, made [test builds](https://github.com/winlibs/winlib-builder/actions/runs/13158628004) (requires https://github.com/winlibs/winlib-builder/tree/curl/winbuild-deprecation; need to check whether this can be used for older cURL versions; need to update to the CMake build chain soonish anyway), and didn't find any further issues when testing locally.

**Comment by cmb69** [comment]
> need to check whether this can be used for older cURL versions

Yep. https://github.com/winlibs/winlib-builder/actions/runs/13160583314. I've pushed the fix.

**Comment by cmb69** [comment]
> need to update to the CMake build chain soonish

A start: https://github.com/winlibs/winlib-builder/pull/40

**Comment by nono303** [comment]
Hi @here,
I did some test with curl 8.12 + php 8.4 and have some issue: 
see. https://github.com/curl/curl/issues/16216#issuecomment-2640549029

**Comment by cmb69** [comment]
@nono303, please don't ping random people. :)

Anyway, I've seen your cURL bug report, but for now that is not an issue for "official" PHP builds, since we don't use c-ares (maybe we should in the future).

**Comment by nono303** [comment]
(Ooops for _here_ 😬)
However, curl 8.12 seems to introduce some {change / regression / bugs}? on my tests, independently of c-ares but most globally related on DNS as RR (experimental) seems to be part of the root cause.
for the moment, not clear at all for me I'll continue digging.

**Comment by nono303** [comment]
sleeping on the issue and coming to the right conclusion 😉
1. There is a curl bug on 8.12 when compiling on windows with `USE_HTTPSRR=OFF` https://github.com/curl/curl/issues/16216#issuecomment-2640338087
2. `curl_setopt_array()` is buggy when setting an option return false https://github.com/php/php-src/issues/17610#issuecomment-2618880731
3. `CURLOPT_DNS_SERVERS` only works when curl is compiled with `ENABLE_ARES=ON` and `ENABLE_THREADED_RESOLVER=OFF` https://github.com/php/php-src/issues/17610#issuecomment-2642367132

This being so _(in point point of view)_, it would be interesting to provide curl lib with c-ares implemented (and thread resolver disabled) to enable `CURLOPT_DNS_SERVERS`

**Comment by cmb69** [comment]
See https://curl.se/mail/lib-2025-02/0017.html

**Comment by Jan-E** [comment]
Curl 8.12.1 will be released on Thursday Feb 13th.

**Comment by cmb69** [comment]
> Curl 8.12.1 will be released on Thursday Feb 13th.

Right, that's [the plan](https://curl.se/mail/lib-2025-02/0027.html).

**Comment by cmb69** [comment]
I've updated to cURL 8.12.1 and [tagged it](https://github.com/winlibs/cURL/releases/tag/libcurl-8.12.1). A [test build](https://github.com/winlibs/winlib-builder/actions/runs/13318941800) for PHP 8.4 didn't show any issues for me locally (besides https://github.com/winlibs/cURL/issues/25#issuecomment-2460582125).

@shivammathur, can you please roll out the new version?

**Comment by shivammathur** [comment]
@cmb69 Done.

**Comment by nono303** [comment]
> I've updated to cURL 8.12.1 and [tagged it](https://github.com/winlibs/cURL/releases/tag/libcurl-8.12.1). A [test build](https://github.com/winlibs/winlib-builder/actions/runs/13318941800) for PHP 8.4 didn't show any issues for me locally 

Same for me for curl 8.12.1 since yesterday.   
all my tests and usages sounds good with php 8.4.4
